### DISCUSSION TITLE
thoughts on archiving tasks

### DISCUSSION THREAD

**Discussion by jaker-dotcom** [GothenburgBitFactory/taskwarrior#2524] [Q&A]
Hey there! 
I have had a wonderful and almost flawless experience with task so far.
However I get lot's of old and deleted tasks when running /searches/ on all tasks that don't matter anymore (but some do).

Can anyone think of a way to prevent this while still preserving the task data for e.g. graphical analysis? Would another file "archive.data" be worthwhile? Or should I just work with a UDA? Anyway I would desire to set rules on when tasks get archived and don't display in search anymore (unless explicitly asked for). 

Or does anyone have an alternative approach? Maybe limiting even the "all" report on tasks modified within the last year?

At this point it comes to mind that I don't know much about what which file holds. Does completed.data also hold deleted tasks? What's the purpose of backlog.data? I couldn't find anything in the docs.
Is there maybe common practice on what to delete every once in a while? Trimming undo.data seems reasonable to me. Anything else? 

Thanks!
Jaker

**Comment by djmitche** [comment]
Yes, completed.data holds all non-pending tasks.  After a `gc`, the numeric id's for tasks are line numbers in `pending.data`, and everything without such an id is in `completed.data`.  The `backlog.data` holds task updates that haven't been sent to the sync server yet.  And `undo.data` as you've guessed holds data required to implement `task undo`.  The data formats aren't documented, to my knowledge, except in the source.  They're known as "TDB2', with the format of each line being called "FF4".

The current task db can get a bit unwieldy (slow and buggy) after a few thousand tasks, so I've typically had to `rm -rf completed.data` about once a year.  Ideally a more compact, indexable on-disk format would allow a much larger set of tasks before performance degraded.  Deleting `undo.data` any time you're sure you don't want to `undo` your latest change is fine, and I do so every time I have to re-initialize a sync replica that has gotten corrupted (so every few weeks).

An approach I have been considering is for a task to have an "expiration date" after which it is permanently removed from the task DB.  That might be stored per-task, or might be configured with some reasonable rules (maybe 1 year after last-modified for a finished task, and 1 month after last-modified for a deleted task).

This would allow the steady-state for a task db to be constant size -- after a few years tasks will expire at the same rate they're being added.  It would also be compatible with more reliable synchronization methods such as Operational Transforms (c.f. #2507).

**Comment by tbabej** [comment]
Hi @jaker-dotcom,

to expand on some of great ideas from @djmitche, you can keep the size of your data files in check even now, with just a bit of manual effort.

    task end.after:2020-01-01 end.before:2021-01-01 export > completed_2020.json  # make a backup with all the tasks completed in 2020
    task end.after:2020-01-01 end.before:2021-01-01 delete  # tasks need to be deleted in order to be purged [1]
    task end.after:2020-01-01 end.before:2021-01-01 purge  # purge permanently removes tasks from the DB

This will remove all completed tasks in year 2020 from your DB. Note that purge command works well only if you're using `task` on a single device.

That said, I've actually never personally experienced bad performance or need to delete my tasks or trim the size of the files. My current task DB has ~2000 tasks after about 6 years of use and loading and filtering the entire DB is still fast:

```
$ time task all
  ... (trimming output of ~1900 tasks)
  - D  703be879   6.6y   6.5y       Personal            2015-01-14 Start doing IGS session                                                                                                                                                                    

1948 tasks

real	0m0.109s
```

Any `task` command on my machine regularly finishes below 200ms, with the most common commands taking 100ms or less.

To answer your original questions, which was more concerned about irrelevant results popping up, you can always use additional filters, for example, assuming your search could be augmented with `entry.after:2021-01-01` to only list tasks that were entered this year:

    task entry.after:2021-01-01 /searchterm/ all

In 2.6.0 (coming soon) you can even make that expression part of your read context, so that you don't have to type it each time. Alternatively you can make it part of your report filter for `all`, which is possible on the currently released `2.5.3`.
